crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -44,6 +44,40 @@ use crate::types::{
 use crate::{Db, FxOrderSet};
 use smallvec::SmallVec;
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum LiteralKind {
+    Int,
+    String,
+    Bytes,
+}
+
+impl<'db> Type<'db> {
+    /// Return `true` if this type can be a supertype of some literals of `kind` and not others.
+    fn splits_literals(self, db: &'db dyn Db, kind: LiteralKind) -> bool {
+        match self {
+            Type::AlwaysFalsy | Type::AlwaysTruthy => true,
```

Comment:
I'm slightly concerned about how much special-casing we're adding for `Type::AlwaysTruthy` and `Type::AlwaysFalsy`, given that conceptually these are really just two protocol types, and we'll be adding a lot more protocol types to our model very soon... but I guess we (I) can cross that bridge when we (I) come to it :-)

Reply:
I think even with the addition of general protocol types, `AlwaysTruthy` and `AlwaysFalsy` will remain "special" in that I don't think any other protocol type can "distinguish" between literals of the same kind?